onnx/defs/math/defs.cc

```diff
@@ -2006,6 +2007,273 @@ ONNX_OPERATOR_SET_SCHEMA(
+        .Input(1, "B", "1 or 2 dimensional data blob", "T2", OpSchema::Single, true, 1, OpSchema::NonDifferentiable)
+        .Input(2, "scales", "quantization scales", "T1", OpSchema::Single, true, 1, OpSchema::NonDifferentiable)
+        .Input(
+            3,
+            "zero_points",
+            "quantization zero points",
+            "T3",
+            OpSchema::Optional,
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .Input(
+            4,
+            "bias",
+            "Bias to add to result. It should have shape [N]",
+            "T1",
+            OpSchema::Optional,
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .Output(
+            0,
+            "Y",
+            "Matrix multiply results. The output tensor has the same rank as the input.",
+            "T1",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .TypeConstraint("T1", {"tensor(float)", "tensor(float16)"},
+                        "Constrain input and output type to float/half_float tensors.")
+        .TypeConstraint("T2", {"tensor(uint8)"}, "Constrain quantized weight types to uint8/int32.")
+        .TypeConstraint("T3", {"tensor(uint8)", "tensor(float16)", "tensor(float)"},
+                        "Constrain quantized zero point types to uint8.int32/float16/float.")
+        .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+          auto k_attr = ctx.getAttribute("K");
+          int64_t K = (k_attr != nullptr) ? k_attr->i() : -1;
+
+          auto n_attr = ctx.getAttribute("N");
+          int64_t N = (n_attr != nullptr) ? n_attr->i() : -1;
+
+          auto bits_attr = ctx.getAttribute("bits");
+          int64_t bits = (bits_attr != nullptr) ? bits_attr->i() : 4;
+
+          auto block_size_attr = ctx.getAttribute("block_size");
+          int64_t block_size = (block_size_attr != nullptr) ? block_size_attr->i() : 128;
+
+          // n & (n-1) is a binary hack to check if n is a power of 2 it will only return zero for pow of 2
+          if (block_size < 16 || ((block_size & (block_size - 1)) != 0)) {
+            fail_type_inference("block_size attribute must be a a power of 2 and not smaller than 16.");
+          }
+
+          auto a_type = ctx.getInputType(0);
+          auto b_type = ctx.getInputType(1);
+          auto scales_type = ctx.getInputType(2);
+          auto y_type = ctx.getOutputType(0);
+          if (nullptr == a_type || nullptr == b_type || nullptr == scales_type || nullptr == y_type ||
+              a_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType ||
+              b_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType ||
+              scales_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType) {
+            fail_type_inference("inputs are expected to have tensor type and output type should not be null.");
+          }
+
```

Comment:
Suggest adding
```cpp
if (!hasNInputShapes(ctx, 3)) {
   return;
}
```
to protect against types of unknown shape/rank. A similar check (eg., `hasInputShape(ctx, 3)`) down below before accessing bias' shape.

Reply:
If I understand correctly you are suggesting that the code return early if we discover no optional inputs.

It is not a bad idea but I don't think that would be correct unless I move the code to propagate the ElemType and output shape. I will look into doing that. 

Am I properly understanding your suggestion?

Reply:
Also: adding unit-tests for the shape-inference would be helpful. You can do that [here](https://github.com/onnx/onnx/blob/main/onnx/test/shape_inference_test.py)

Reply:
I added the code as suggested as well calls to `hasinputshape(ctx, 3/4)` for the optional inputs `zero_points` and `bias`.

I also added `MatMulNBits` to the shape-inference unit test thanks for pointing that out to me.